test/dist/dist_test.go

```diff
@@ -74,24 +76,191 @@ func TestPullManifest(t *testing.T) {
 	regAuthCtx.Scope.RemoteName = remoteName
 	regAuthCtx.Scope.Actions = "pull"
 
+	if err := regAuthCtx.PrepareAuth(indexServer); err != nil {
+		return "", fmt.Errorf("failed to prepare auth to %s for %s: %v", indexServer, reqPath, err)
+	}
+
+	inputURL := "https://" + indexServer + "/v2/" + reqPath
+
+	res, err := regAuthCtx.GetResponse(inputURL, "HEAD", nil, []int{http.StatusOK})
+	if err != nil {
+		return "", fmt.Errorf("got an unexpected reply: %v", err)
+	}
+
+	return res.Header.Get(distp.ContentDigest), nil
+}
+
+func testUploadLayer(t *testing.T) {
+	regAuthCtx := auth.NewRegAuthContext()
+	regURL := regAuthCtx.RegURL
+	indexServer := auth.GetIndexServer(regURL)
+
+	remoteName := filepath.Join(auth.DefaultRepoPrefix, testImageName)
+	reqPath := filepath.Join(remoteName, "blobs/uploads", testRefName)
+
+	regAuthCtx.Scope.RemoteName = remoteName
+	regAuthCtx.Scope.Actions = "push"
+
+	if err := regAuthCtx.PrepareAuth(indexServer); err != nil {
+		t.Fatalf("failed to prepare auth to %s for %s: %v", indexServer, reqPath, err)
+	}
+
+	// 1. POST
+	// Send a POST request without any body specified.
+	postURL := "https://" + indexServer + "/v2/" + reqPath
+
+	res, err := regAuthCtx.GetResponse(postURL, "POST", nil, []int{http.StatusOK, http.StatusAccepted})
+	if err != nil {
+		t.Fatalf("got an unexpected reply: %v", err)
+	}
+
+	uuid := res.Header.Get(distp.UploadUuidKey)
+
+	// 2. PATCH
+	// Generate a 100-byte blob of a randomly generated string.
+	// Send a PATCH request with the blob.
+	blob := image.GenRandomBlob(100)
+
+	if _, err := regAuthCtx.GetResponse(postURL, "PATCH", strings.NewReader(blob),
+		[]int{http.StatusOK, http.StatusAccepted}); err != nil {
+		t.Fatalf("got an unexpected reply: %v", err)
+	}
+
+	// 3. PUT
+	// Generate a blob's digest, generated as a sha256 checksum of the blob.
+	// Send a PUT request with a "digest=..." option appended to its URL.
+	digest := image.GetHash(blob)
+	putURL := "https://" + indexServer + "/v2/" + reqPath + "/" + uuid + "?digest=" + digest
```

Comment:
i'm wondering if a next step could be to have a golang package much like https://github.com/docker/distribution/blob/master/registry/api/v2/descriptors.go#L237 that clients could import and this test suite could also use to validate against.